### PR TITLE
New header src/macro_helpers.h that supports preprocessor-macros

### PR DESCRIPTION
**Pull request by noseglasses**
Working on my todo list, I came back to an old topic, where users reported problems when their
keymap definitions where having too many macro arguments or if they added spurious commas, see

https://community.keyboard.io/t/compilation-errors-with-complaints-about-keymap-stacked/538

In the topic I proposed a fix that is based on a static assertion. This fix is the main part of this PR.

It adds a macro `RESTRICT_ARGS_COUNT` that can be used to check the number of arguments in macro functions that are passed a lot of arguments like `KEYMAP_STACKED`.

The macro can be added to the definitions of macros `KEYMAP(...)` and `KEYMAP_STACKED(...)` in the hardware modules, e.g. Kaleidoscope-Hardware-Model01. 

```cpp
#define KEYMAP_STACKED(                                                 \
               r0c0, r0c1, r0c2, r0c3, r0c4, r0c5, r0c6,                \
               r1c0, r1c1, r1c2, r1c3, r1c4, r1c5, r1c6,                \
               r2c0, r2c1, r2c2, r2c3, r2c4, r2c5,                      \
               r3c0, r3c1, r3c2, r3c3, r3c4, r3c5, r2c6,                \
               r0c7, r1c7, r2c7, r3c7,                                  \
               r3c6,                                                    \
                                                                        \
               r0c9,  r0c10, r0c11, r0c12, r0c13, r0c14, r0c15,         \
               r1c9,  r1c10, r1c11, r1c12, r1c13, r1c14, r1c15,         \
                      r2c10, r2c11, r2c12, r2c13, r2c14, r2c15,         \
               r2c9,  r3c10, r3c11, r3c12, r3c13, r3c14, r3c15,         \
               r3c8,  r2c8,  r1c8, r0c8,                                \
               r3c9, ...)                                               \
  {                                                                     \
    {r0c0, r0c1, r0c2, r0c3, r0c4, r0c5, r0c6, r0c7, r0c8, r0c9, r0c10, r0c11, r0c12, r0c13, r0c14, r0c15}, \
    {r1c0, r1c1, r1c2, r1c3, r1c4, r1c5, r1c6, r1c7, r1c8, r1c9, r1c10, r1c11, r1c12, r1c13, r1c14, r1c15}, \
    {r2c0, r2c1, r2c2, r2c3, r2c4, r2c5, r2c6, r2c7, r2c8, r2c9, r2c10, r2c11, r2c12, r2c13, r2c14, r2c15}, \
    {r3c0, r3c1, r3c2, r3c3, r3c4, r3c5, r3c6, r3c7, r3c8, r3c9, r3c10, r3c11, r3c12, r3c13, r3c14, RESTRICT_ARGS_COUNT((r3c15), 64, KEYMAP_STACKED, ##__VA_ARGS__)}, \
  }
```

PRs for the existing hardware definitions follow.

A test of the new argument check can e.g. be performed when adding an additional argument at the front of an invocation of `KEYMAP_STACKED(...)`. Please note that `ShiftToLayer(FUNCTION)` is the last intended argument in the list that is thus shifted. 

The compiler will output
```
***************************************************************
******************** READ THIS CAREFULLY! *********************
***************************************************************

File: .../hardware/keyboardio/avr/libraries/Model01-Firmware/Model01-Firmware.ino
Line: 146

Strange arguments encountered in call to KEYMAP_STACKED.

Please make sure to pass exactly 64 macro arguments to
KEYMAP_STACKED. Also make sure that there are no dangling
commas at the end of the argument list. This is the superfluous
part at the end of the macro arguments list.

'ShiftToLayer(FUNCTION)'

***************************************************************
***************************************************************
***************************************************************
```
When an additional argument is added to end of the list, e.g. named `foo`, the compiler error message is
```
***************************************************************
******************** READ THIS CAREFULLY! *********************
***************************************************************

File: .../hardware/keyboardio/avr/libraries/Model01-Firmware/Model01-Firmware.ino
Line: 146

Strange arguments encountered in call to KEYMAP_STACKED.

Please make sure to pass exactly 64 macro arguments to
KEYMAP_STACKED. Also make sure that there are no dangling
commas at the end of the argument list. This is the superfluous
part at the end of the macro arguments list.

'foo'

***************************************************************
***************************************************************
***************************************************************
```

A single superfluous comma at the end of the argument list is tolerated. This is because it is incredibly complex to catch this corner case with preprocessor macro processing. On the other hand, a dangling comma at the end of the initialization list is simply ignored by the compiler, as long as all other 64 macro parameters are well defined. Thus, it is safe to ignore this corner case.

More than one superfluous commas (here 2) results in the compiler error
```
***************************************************************
******************** READ THIS CAREFULLY! *********************
***************************************************************

File: .../hardware/keyboardio/avr/libraries/Model01-Firmware/Model01-Firmware.ino
Line: 146

Strange arguments encountered in call to KEYMAP_STACKED.

Please make sure to pass exactly 64 macro arguments to
KEYMAP_STACKED. Also make sure that there are no dangling
commas at the end of the argument list. This is the superfluous
part at the end of the macro arguments list.

','

***************************************************************
***************************************************************
***************************************************************
```

Moreover, this PR adds the auxiliary macros `__NN__` and `__NL__` that aid the use
of multiline macros when reading pre-processed code (see comments in the code).

### CODE REVIEW

**Comment by noseglasses** [comment]
Check out

https://community.keyboard.io/t/error-macro-keymap-stacked-requires-64-arguments-but-only-62-given-help-please/1010

to see why this PR might be useful.

File: src/macro_helpers.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#define __NN__
+
+// Some auxiliary macros
+//
+#define __STRINGIZE(S) #S
+#define STRINGIZE(S) __STRINGIZE(S)
+
+// Allow for the creation of verbose messages in static_asserts
+//
+#define VEROSE_STATIC_ASSERT_HEADER                                            \
```
Comment (gedankenexperimenter) [inline src/macro_helpers.h]: Maybe you want to correct the spelling: s/VEROSE/VERBOSE/

**Comment by noseglasses** [comment]
> Ugh, this fell off my radar, sorry!

I am happy to see that there's a lot of progress in many parts of the firmware with numerous PRs and issues. It is perfectly understandable that not everything can be on the radar at the same time.
I wonder how it is going to be when people have received their MP2 and MP3 keyboards. I see a veritable stress test approaching your radar ;-)

**Comment by algernon** [comment]
I really, really, really like this, so unless @obra objects, I'll merge it over the weekend.

**Comment by obra** [comment]
Thanks. Applied.

**Pull request merged**